File: rxbinding-support-v4/src/main/java/com/jakewharton/rxbinding2/support/v4/widget/SlidingPaneLayoutSlideObservable.java
package com.jakewharton.rxbinding2.support.v4.widget;

import android.support.v4.widget.SlidingPaneLayout;
import android.view.View;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static io.reactivex.android.MainThreadDisposable.verifyMainThread;

final class SlidingPaneLayoutSlideObservable extends Observable<Float> {
  private final SlidingPaneLayout view;

  SlidingPaneLayoutSlideObservable(SlidingPaneLayout view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Float> observer) {
    verifyMainThread();
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setPanelSlideListener(listener);
  }

  static final class Listener extends MainThreadDisposable
      implements SlidingPaneLayout.PanelSlideListener {
    private final SlidingPaneLayout view;
    private final Observer<? super Float> observer;

    Listener(SlidingPaneLayout view, Observer<? super Float> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onPanelSlide(View panel, float slideOffset) {
      if (!isDisposed()) {
        observer.onNext(slideOffset);
      }
    }

    @Override public void onPanelOpened(View panel) {

    }

    @Override public void onPanelClosed(View panel) {

    }

    @Override protected void onDispose() {
      view.setPanelSlideListener(null);
    }
  }
}

File: rxbinding-slidingpanelayout/src/main/java/com/jakewharton/rxbinding3/slidingpanelayout/SlidingPaneLayoutSlideObservable.kt
@file:JvmName("RxSlidingPaneLayout")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.slidingpanelayout

import android.view.View
import androidx.annotation.CheckResult
import androidx.slidingpanelayout.widget.SlidingPaneLayout
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

import com.jakewharton.rxbinding3.internal.checkMainThread

/**
 * Create an observable of the slide offset of the pane of `view`
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Warning:* The created observable uses [SlidingPaneLayout.setPanelSlideListener]
 * to observe dismiss change. Only one observable can be used for a view at a time.
 */
@CheckResult
fun SlidingPaneLayout.panelSlides(): Observable<Float> {
  return SlidingPaneLayoutSlideObservable(this)
}

private class SlidingPaneLayoutSlideObservable(
  private val view: SlidingPaneLayout
) : Observable<Float>() {

  override fun subscribeActual(observer: Observer<in Float>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.setPanelSlideListener(listener)
  }

  private class Listener(
    private val view: SlidingPaneLayout,
    private val observer: Observer<in Float>
  ) : MainThreadDisposable(), SlidingPaneLayout.PanelSlideListener {

    override fun onPanelSlide(panel: View, slideOffset: Float) {
      if (!isDisposed) {
        observer.onNext(slideOffset)
      }
    }

    override fun onPanelOpened(panel: View) {
    }

    override fun onPanelClosed(panel: View) {
    }

    override fun onDispose() {
      view.setPanelSlideListener(null)
    }
  }
}
